File: platform-console/apps/raap-admin/src/api/core/rlhf.ts
import { requestClient } from '#/api/request';

/** RLHF 反馈模型 */
export namespace RLHFApi {
  export interface RLHFFeedback {
    id: number;
    job_id: string;
    sub_job_id: string;
    content_id: string;
    trace_id?: string;
    title?: string;
    content?: string;
    modified_title?: string;
    modified_content?: string;
    ge_expert_code?: string;
    model_code?: string;
    like_status: -1 | 0 | 1;
    like_reason?: string;
    like_user_name?: string;
    adopt_status: -1 | 0 | 1 | 2;
    adopt_reason?: string;
    discard_reason_type?: string;
    improvement_suggestion?: string;
    content_score: number;
    model_score: number;
    issue_tag_ids?: number[];
    custom_issue_tags?: string[];
    is_locked: 0 | 1;
    lock_user_name?: string;
    lock_expire_time?: string;
    review_status:
      | 'COMPLETED'
      | 'IN_INSPECTION'
      | 'IN_PROGRESS'
      | 'INSPECTION_FAILED'
      | 'INSPECTION_PASSED'
      | 'PENDING';
    created_at: string;
    updated_at: string;
    // 审核人信息
    review_user_id?: string;
    review_user_name?: string;
    review_time?: string;
    // 抽检结果相关 (附属字段)
    inspection_comment?: string;
    inspection_user_id?: string;
    inspection_user_name?: string;
    inspection_time?: string;
    // 上下文变量 (从 Content 表关联获取)
    context_list?: Record<string, string>;
    // 划词评论
    annotations?: any[];
  }

  export interface RLHFIssueTag {
    id: number;
    tag_code: string;
    tag_name: string;
    tag_category?: string;
    description?: string;
    enabled: 0 | 1;
    sort_order: number;
  }

  export interface RLHFStatsSummary {
    total_count: number;
    pending_count: number;
    completed_count: number;
    like_rate: number;
    adopt_rate: number;
    avg_content_score: number;
  }
}

/** 获取内容列表 */
export function getRLHFListApi(params: any) {
  return requestClient.get<any>('/v1/rlhf/contents', { params });
}

/** 获取内容详情 */
export function getRLHFDetailApi(id: number) {
  return requestClient.get<RLHFApi.RLHFFeedback>(`/v1/rlhf/contents/${id}`);
}

/** 随机获取待审核内容 */
export function getRandomRLHFContentsApi(count = 1) {
  return requestClient.get<RLHFApi.RLHFFeedback[]>('/v1/rlhf/contents/random', {
    params: { count },
  });
}

/** 锁定内容 */
export function lockContentApi(id: number) {
  return requestClient.post(`/v1/rlhf/contents/${id}/lock`);
}

/** 解锁内容 */
export function unlockContentApi(id: number) {
  return requestClient.post(`/v1/rlhf/contents/${id}/unlock`);
}

/** 解锁当前用户锁定的所有内容 */
export function unlockAllMyContentsApi() {
  return requestClient.post<{ success: boolean; unlocked_count: number }>(
    '/v1/rlhf/contents/unlock-all',
  );
}

/** 批量锁定内容 */
export function batchLockContentsApi(ids: number[]) {
  return requestClient.post<{
    auto_unlocked_count?: number;
    auto_unlocked_ids?: number[];
    failed_count: number;
    failed_ids: number[];
    message?: string;
    success_count: number;
    success_ids: number[];
  }>('/v1/rlhf/contents/batch-lock', { ids });
}

/** 批量解锁内容 */
export function batchUnlockContentsApi(ids: number[]) {
  return requestClient.post<{
    failed_count: number;
    failed_ids: number[];
    success_count: number;
    success_ids: number[];
  }>('/v1/rlhf/contents/batch-unlock', { ids });
}

/** 心跳续锁 - 延长锁定时间 */
export function renewLocksApi(ids: number[]) {
  return requestClient.post<{
    new_expire_time: string;
    renewed_count: number;
    success: boolean;
  }>('/v1/rlhf/contents/renew-locks', { ids });
}

/** 喜欢/不喜欢 */
export function likeContentApi(
  id: number,
  data: { improvement_suggestion?: string; reason: string; status: -1 | 1 },
) {
  return requestClient.post(`/v1/rlhf/contents/${id}/like`, data);
}

/** 采纳/不采纳/废弃 */
export function adoptContentApi(
  id: number,
  data: {
    discard_reason_type?: string;
    improvement_suggestion?: string;
    reason: string;
    status: -1 | 1 | 2;
  },
) {
  return requestClient.post(`/v1/rlhf/contents/${id}/adopt`, data);
}

/** 评分 */
export function scoreContentApi(
  id: number,
  data: {
    content_score: number;
    custom_issue_tags?: string[];
    issue_tag_ids?: number[];
    model_score: number;
    modified_content?: string;
    modified_title?: string;
  },
) {
  return requestClient.post(`/v1/rlhf/contents/${id}/score`, data);
}

/** 抽检 */
export function inspectionContentApi(
  id: number,
  data: {
    comment?: string;
    issue_tag_names?: string[];
    result: 'FAILED' | 'PASSED';
  },
) {
  return requestClient.post<RLHFApi.RLHFFeedback>(
    `/v1/rlhf/contents/${id}/inspection`,
    data,
  );
}

/** AI 建议标签 */
export function suggestRLHFTagsApi(id: number, data: { comment?: string }) {
  return requestClient.post<{ tags: string[] }>(
    `/v1/rlhf/contents/${id}/suggest-tags`,
    data,
  );
}

/** AI 总结意见 - 根据原文和划词评论生成修改意见 */
export function summarizeRLHFCommentApi(
  id: number,
  data: { model_code?: string } = {},
) {
  return requestClient.post<{ comment: string }>(
    `/v1/rlhf/contents/${id}/summarize-comment`,
    data,
  );
}

/** 更新内容（包括 AI 意见和标签） */
export function updateContentApi(
  id: number,
  data: {
    annotations?: unknown[];
    improvement_suggestion?: string;
    issue_tag_names?: string[];
    modified_content?: string;
    modified_title?: string;
  },
) {
  return requestClient.put(`/v1/rlhf/contents/${id}`, data);
}

/** 原文精修 - 保存精修后的标题和内容 */
export function refineContentApi(
  id: number,
  data: {
    refined_content?: string;
    refined_title?: string;
  },
) {
  return requestClient.post<RLHFApi.RLHFFeedback>(
    `/v1/rlhf/contents/${id}/refine`,
    data,
  );
}

/** 获取历史 */
export function getRLHFHistoryApi(id: number) {
  return requestClient.get<any[]>(`/v1/rlhf/contents/${id}/history`);
}

/** 获取标签列表 */
export function getIssueTagsApi() {
  return requestClient.get<RLHFApi.RLHFIssueTag[]>('/v1/rlhf/issue-tags');
}

/** 创建标签 */
export function createIssueTagApi(data: any) {
  return requestClient.post('/v1/rlhf/issue-tags', data);
}

/** 更新标签 */
export function updateIssueTagApi(id: number, data: any) {
  return requestClient.put(`/v1/rlhf/issue-tags/${id}`, data);
}

/** 删除标签 */
export function deleteIssueTagApi(id: number) {
  return requestClient.delete(`/v1/rlhf/issue-tags/${id}`);
}

/** 获取统计摘要 */
export function getRLHFStatsSummaryApi() {
  return requestClient.get<RLHFApi.RLHFStatsSummary>('/v1/rlhf/stats/summary');
}

/** 获取状态选项 */
export function getReviewStatusOptionsApi() {
  return requestClient.get<Array<{ label: string; value: string }>>(
    '/v1/rlhf/review-status-options',
  );
}

/** 获取审核人列表（用于下拉筛选） */
export function getReviewersApi() {
  return requestClient.get<Array<{ label: string; value: string }>>(
    '/v1/rlhf/reviewers',
  );
}

/** 更新审核状态（喜欢/不喜欢） */
export function updateReviewStatusApi(
  id: number,
  data: { comment?: string; issue_tag_names?: string[]; review_status: string },
) {
  return requestClient.put(`/v1/rlhf/contents/${id}/review-status`, data);
}

File: platform-server/app/schemas/rlhf.py
from typing import Optional, List, Dict, Any, Union
from datetime import datetime
from pydantic import BaseModel, Field, ConfigDict

# --- Issue Tag Schemas ---

class RLHFIssueTagBase(BaseModel):
    tag_code: str = Field(..., description="标签编码")
    tag_name: str = Field(..., description="标签名称")
    tag_category: Optional[str] = Field(None, description="标签分类：CONTENT/MODEL/BRAND/COMPLIANCE/OTHER")
    description: Optional[str] = None
    enabled: int = Field(1, description="是否启用：0禁用 1启用")
    sort_order: int = Field(0, description="排序")

    model_config = ConfigDict(from_attributes=True, protected_namespaces=())

class RLHFIssueTagCreate(RLHFIssueTagBase):
    pass

class RLHFIssueTagUpdate(BaseModel):
    tag_name: Optional[str] = None
    tag_category: Optional[str] = None
    description: Optional[str] = None
    enabled: Optional[int] = None
    sort_order: Optional[int] = None

class RLHFIssueTagOut(RLHFIssueTagBase):
    id: int
    use_count: int
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True, protected_namespaces=())

# --- Operation History Schemas ---

class RLHFOperationHistoryOut(BaseModel):
    id: int
    feedback_id: int
    operation_type: str
    before_value: Optional[Dict[str, Any]] = None
    after_value: Optional[Dict[str, Any]] = None
    reason: Optional[str] = None
    improvement_suggestion: Optional[str] = None
    operator_id: str
    operator_name: Optional[str] = None
    operation_time: datetime

    model_config = ConfigDict(from_attributes=True, protected_namespaces=())

# --- Feedback Schemas ---

class RLHFFeedbackBase(BaseModel):
    title: Optional[str] = None
    content: Optional[str] = None
    modified_title: Optional[str] = None
    modified_content: Optional[str] = None
    
    ge_expert_code: Optional[str] = None
    ag_expert_codes: Optional[List[str]] = None
    model_code: Optional[str] = None
    
    like_status: int = 0
    like_reason: Optional[str] = None
    adopt_status: int = 0
    adopt_reason: Optional[str] = None
    discard_reason_type: Optional[str] = None
    discard_comment: Optional[str] = None
    improvement_suggestion: Optional[str] = None
    
    content_score: float = 0
    model_score: float = 0
    
    issue_tag_ids: Optional[List[int]] = None
    custom_issue_tags: Optional[List[str]] = None
    
    annotations: Optional[List[Dict[str, Any]]] = None
    
    review_status: str = "PENDING"
    inspection_comment: Optional[str] = None
    inspection_user_id: Optional[str] = None
    inspection_user_name: Optional[str] = None
    inspection_time: Optional[datetime] = None
    
    model_config = ConfigDict(protected_namespaces=())

class RLHFFeedbackCreate(BaseModel):
    job_id: str
    sub_job_id: str
    content_id: str
    trace_id: Optional[str] = None
    title: Optional[str] = None
    content: Optional[str] = None
    ge_expert_code: Optional[str] = None
    model_code: Optional[str] = None
    review_status: str = "PENDING"
    created_by: Optional[str] = None

    model_config = ConfigDict(protected_namespaces=())

class RLHFFeedbackUpdate(BaseModel):
    modified_title: Optional[str] = None
    modified_content: Optional[str] = None
    annotations: Optional[List[Dict[str, Any]]] = None
    # AI 意见和标签（支持自动保存）
    improvement_suggestion: Optional[str] = Field(None, description="AI 生成的修改意见")
    issue_tag_names: Optional[List[str]] = Field(None, description="问题标签名称列表")

    model_config = ConfigDict(protected_namespaces=())

class RLHFFeedbackOut(RLHFFeedbackBase):
    id: int
    job_id: str
    sub_job_id: str
    content_id: str
    trace_id: Optional[str] = None
    
    modify_count: int = 0
    
    like_user_id: Optional[str] = None
    like_user_name: Optional[str] = None
    like_time: Optional[datetime] = None
    
    adopt_user_id: Optional[str] = None
    adopt_user_name: Optional[str] = None
    adopt_time: Optional[datetime] = None
    
    is_locked: int = 0
    lock_user_id: Optional[str] = None
    lock_user_name: Optional[str] = None
    lock_time: Optional[datetime] = None
    lock_expire_time: Optional[datetime] = None
    
    created_at: datetime
    updated_at: datetime
    
    # 审核人信息
    review_user_id: Optional[str] = None
    review_user_name: Optional[str] = None
    review_time: Optional[datetime] = None
    
    # 抽检相关
    inspection_status: Optional[str] = "PENDING"
    inspection_result: Optional[str] = None
    inspection_comment: Optional[str] = None
    inspection_user_id: Optional[str] = None
    inspection_user_name: Optional[str] = None
    inspection_time: Optional[datetime] = None
    
    # 上下文变量 (从 Content 表关联获取)
    context_list: Optional[Dict[str, Any]] = None
    
    model_config = ConfigDict(from_attributes=True, protected_namespaces=())

# --- Operation Request Schemas ---

class RLHFLikeRequest(BaseModel):
    status: int = Field(..., description="1: 喜欢, -1: 不喜欢")
    reason: str = Field(..., min_length=30, description="原因（≥30字）")
    improvement_suggestion: Optional[str] = Field(None, description="改进建议（不喜欢时必填）")

class RLHFAdoptRequest(BaseModel):
    status: int = Field(..., description="1: 采纳, -1: 不采纳, 2: 废弃")
    reason: str = Field(..., min_length=30, description="原因（≥30字）")
    discard_reason_type: Optional[str] = None
    improvement_suggestion: Optional[str] = Field(None, description="改进建议（不采纳时必填）")

class RLHFScoreRequest(BaseModel):
    content_score: float = Field(..., ge=1, le=10, description="内容评分(1-10)")
    model_score: float = Field(..., ge=1, le=10, description="模型评分(1-10)")
    issue_tag_ids: Optional[List[int]] = None
    custom_issue_tags: Optional[List[str]] = None
    modified_title: Optional[str] = None
    modified_content: Optional[str] = None
    
    model_config = ConfigDict(protected_namespaces=())

class RLHFTagRequest(BaseModel):
    issue_tag_ids: Optional[List[int]] = None
    custom_issue_tags: Optional[List[str]] = None

class RLHFInspectionRequest(BaseModel):
    result: str = Field(..., description="PASSED 或 FAILED")
    comment: Optional[str] = Field(None, description="抽检意见")
    issue_tag_names: Optional[List[str]] = Field(None, description="问题标签名称列表（可选，支持自动新增）")

    model_config = ConfigDict(protected_namespaces=())

class RLHFSummaryRequest(BaseModel):
    comment: Optional[str] = Field(None, description="抽检人填写的初步修改意见")

    model_config = ConfigDict(protected_namespaces=())

class RLHFSummaryResponse(BaseModel):
    tags: List[str] = Field(..., description="AI 建议的问题标签")

    model_config = ConfigDict(protected_namespaces=())


class RLHFSummarizeCommentRequest(BaseModel):
    """AI 总结意见请求"""
    model_code: str = Field("gpt-4o", description="使用的模型编码")

    model_config = ConfigDict(protected_namespaces=())


class RLHFSummarizeCommentResponse(BaseModel):
    """AI 总结意见响应"""
    comment: str = Field(..., description="AI 生成的修改意见")

    model_config = ConfigDict(protected_namespaces=())

class RLHFLockResponse(BaseModel):
    success: bool
    message: str
    lock_expire_time: Optional[datetime] = None


# --- Stats Schemas ---

class RLHFStatsSummary(BaseModel):
    total_count: int
    pending_count: int
    completed_count: int
    like_rate: float
    adopt_rate: float
    avg_content_score: float
    
class RLHFReviewerStats(BaseModel):
    reviewer_id: str
    reviewer_name: str
    total_count: int
    like_count: int
    adopt_count: int
    avg_score: float

class RLHFDailyStatsOut(BaseModel):
    stat_date: str
    total_count: int
    like_rate: float
    adopt_rate: float
    avg_content_score: float
    
    model_config = ConfigDict(from_attributes=True, protected_namespaces=())

